File: toys/algos.2/epi/hanoi.c
#include <stdio.h>

void Hanoi(int n, int from, int to, int use) {
    if (n > 0) {
        Hanoi(n - 1, from, use, to);
        printf("Move from peg %d to %d\n", from, to);
        Hanoi(n - 1, use, to, from);
    }
}

int main()
{
    Hanoi(4, 1, 3, 2);
    return 0;
}

File: toys/misc/perl/hanoi.pl
use warnings;
use strict;

# Towers of Hanoi
# Perl version (5.8.0)
# Ported from Java

my $numdisks = 0;

print "Number of disks? ";
chomp( $numdisks = <STDIN> );

print "The moves are:\n\n";
movedisks( $numdisks, 'A', 'B', 'C' );

sub movedisks {

  my( $num, $from, $to, $aux ) = @_;

  if( $num == 1 ) {
    print "Move disk $num from $from to $to\n";
  }

  else {
    movedisks( $num-1, $from, $aux, $to );
    print "Move disk $num from $from to $to\n";
    movedisks( $num-1, $aux, $to, $from );
  }
}

